src/v/ssx/thread_worker.h

```diff
@@ -110,73 +153,111 @@ class thread_worker {
 
     ss::future<> stop() {
         co_await _gate.close();
-        _ready.signal(RUNNABLE_STOP);
         if (_worker.joinable()) {
+            auto task = impl::stop_task();
+            vassert(
+              _queue.push(&task),
+              "expected to be able to push a task onto the queue");
+            _ready.signal(1);
+            co_await task.get_future();
             _worker.join();
         }
     }
 
     template<typename Func>
-    auto submit(Func func) {
+    auto submit(Func func) ->
+      typename ss::futurize<std::invoke_result_t<Func>>::type {
         auto gh = _gate.hold();
-        return _mutex.get_units().then(
-          [this, func{std::move(func)}, gh{std::move(gh)}](auto units) mutable {
-              vassert(_task == nullptr, "Only one task supported at a time");
-              auto task = std::make_unique<impl::worker_task<Func>>(
-                std::move(func));
-              auto f = task->get_future().finally(
-                [this, gh{std::move(gh)}, units{std::move(units)}] {
-                    _task.reset();
-                });
-              _task = std::move(task);
-              _ready.signal(RUNNABLE_READY);
-              return f;
-          });
+        auto units = co_await ss::get_units(_semaphore, 1);
+        auto task = impl::worker_task<Func>(std::move(func));
+        vassert(
+          _queue.push(&task),
+          "expected to be able to push a task onto the queue");
+        _ready.signal(1);
+        co_return co_await task.get_future();
     }
 
 private:
     void configure_thread() {
-        ::pthread_setname_np(::pthread_self(), "ssx::thread_worker");
+        auto name = ss::format("{}_{}", _config->name.c_str(), _shard_id);
```

Comment:
Is the `c_str()` needed? Seems to require a `strlen` rather than allowing format to know about the length.

Reply:
Nope it's vestigial from when I wasn't adding the shard_id here.